lgc/builder/BuilderImpl.cpp

```diff
@@ -388,90 +380,133 @@ static Value *traceNonUniformIndex(Value *nonUniformVal) {
     }
   }
 
-  SmallVector<Value *, 2> worklist;
-  Value *base = load->getOperand(0);
-  Value *index = nullptr;
+  auto getSize = [](Value *value) -> uint64_t {
+    uint64_t size = value->getType()->getPrimitiveSizeInBits().getFixedValue();
+    return size ? size : std::numeric_limits<uint64_t>::max();
+  };
 
-  // Loop until a descriptor table reference or unexpected operation is reached.
-  // In the worst case this may visit all instructions in a function.
-  for (;;) {
-    if (auto bitcast = dyn_cast<BitCastInst>(base)) {
-      base = bitcast->getOperand(0);
-      continue;
+  uint64_t nonUniformValSize = getSize(nonUniformVal);
+
+  // Loop until all nonUniforms have been found to be uniform or a heuristic abort criterion has been reached.
+  Value *candidateIndex = nullptr;
+  SmallVector<Instruction *, 2> nonUniforms;
+  nonUniforms.push_back(load);
+
+  do {
+    Instruction *current = nonUniforms.pop_back_val();
+
+    // Immediately replace the current nonUniformVal by a strictly smaller one if possible.
+    if (!candidateIndex && nonUniforms.empty() && current != nonUniformVal) {
+      uint64_t size = getSize(current);
+      if (size < nonUniformValSize) {
+        nonUniformVal = current;
+        nonUniformValSize = size;
+      }
     }
-    if (auto gep = dyn_cast<GetElementPtrInst>(base)) {
-      if (gep->hasAllConstantIndices()) {
-        base = gep->getPointerOperand();
-        continue;
+
+    // See if we can propagate the search further.
+    auto propagate = [&](Value *value) -> bool {
```

Comment:
Purely stylistic, but presumably declaration doesn't need to be in the loop?

Reply:
Thank you for taking a look. My thinking was to have it close to its uses, but I don't think it really matters much. I've moved it out of the loop.